devito/ir/support/utils.py

```diff
@@ -69,11 +69,16 @@ def detect_flow_directions(exprs):
                 continue
             dimensions = [d for d in w.aindices if d is not None]
             for d in dimensions:
+                if d.is_Sub:
+                    rootd = d.parent
+                else:
+                    rootd = d
+
                 try:
-                    if w.distance(r, d) > 0:
+                    if w.distance(r, rootd) > 0:
                         mapper[d].add(Forward)
                         break
-                    elif w.distance(r, d) < 0:
+                    elif w.distance(r, rootd) < 0:
```

Comment:
wait, don't we need ``rootd`` in place of ``d`` a few lines below as well?

Reply:
Where were you thinking?  I confess I don't have a great handle on what everything in here does, just yet, so my instinct was to keep the changes to a minimum.  For this bug at least, it really just seemed to be the distance() call that was failing.

Reply:
oh, OK, sorry, I should have provided more info. I meant lines 82, 85, 88

nitpicking: for the convention used throughout Devito, can we use ``dim`` rather than ``rootd``?

Reply:
and in fact, now that I see it, line 91 too:
```
for d in dimensions[dimensions.index(dim) + 1:]:
```

Reply:
nope, this won't work because ``dim`` isn't in ``dimensions``. I think we have to it this way:
```
dimensions = [d for d in w.aindices if d is not None]
dimensions = [d.parent if d.is_Sub else d for d in dimensions]
...
```
then you can leave all the rest as in ``master``

Reply:
With just that change, it doesn't get the direction right anymore.  The subdimension doesn't get the direction assigned to it - and ends up being caught by this:

```
    # Add in any encountered Dimension
    mapper.update({d: {Any} for d in flatten(i.aindices for i in reads + writes)
                   if d is not None and d not in mapper})
```